src/Compilers/Core/Portable/Compilation/DeterministicKey.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract class SyntaxTreeKey
```

Comment:
What other derivations of this type were you expecting to write? My intuition was to use a `readonly struct` or `sealed class` here but seems like you had other scenarios in mind that I'm missing.

Reply:
In the IDE i implement this by sitting on top of our Document abstraction.  So GetText forwards to our own impl of getting the text.  It could a readonly-struct, but it would need to be passed a Delegate to do the work.  This seemed cleaner (and in in line with all our 'XXXProvider' types.